src/main/java/com/xpanxion/java/springboot/da1/demo/service/student8/GymService8.java

```diff
@@ -48,32 +51,56 @@ public Gym8 createGym(Gym8 gym) {
     }
 
     public Member8 addMember(Member8 member, Integer gymId) {
-        member.setGymId(gymId);
+        Gym8 gym = gymRepository8.getFindById(gymId);
+        if (gym == null) {
+            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Gym not found");
+        }
+        member.setGym(gym);
         return memberRepository8.save(member);
     }
 
-    public Timestamp8 memberCheckIn(Integer memberId, String checkInTime) {
+    public Member8 findMember(int memberId) {
+        return memberRepository8.getById(memberId);
+    }
+
+    public Timestamp8 memberCheckIn(int memberId, Timestamp checkInTime, CheckInOutType8 checkType) {
         if (memberRepository8.findById(memberId).isPresent()) {
-            var timestamp = new Timestamp8(memberId, checkInTime);
+            Member8 member = findMember(memberId);
+            MemberHistory8 memberRecord = new MemberHistory8(member, checkInTime, checkType);
+            memberHistoryRepository8.save(memberRecord);
+            Timestamp8 timestamp = new Timestamp8(member, checkInTime);
             return timestampRepository8.save(timestamp);
         }
         else {
             throw new ResponseStatusException(HttpStatus.NOT_FOUND);
         }
     }
 
-    public Timestamp8 memberCheckOut(Integer memberId, String checkOutTime) {
-        if(memberRepository8.findById(memberId).isPresent()){
```

Comment:
Everything looks great.  One possible improvement for readability, instead of `isPresent()` all the cool kids are doing this:

`orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Member not found."));`

Reply:
I created a helper..
```
    private Member0 getMemberOrThrow(int memberId) {
        var retval= memberRepository0.findById(memberId).orElseThrow(() ->
                new ResponseStatusException(HttpStatus.NOT_FOUND, "Member not found."));
        return retval;
    }
```

Reply:
thank you! I added it and refactored GymService8 in the latest commit